File: 2012_2013_sujet_4_problemes_divers/TP4_2012_CP.m
%% 1
A=[0 1/3 2/3;1/3 0 2/3;2/3 1/3 0];
P=[0.35 0.25 0.4];
P*A
%%
p0=[0.05 0.9 0.05];
p5=p0*A^5
p10=p0*A^10
p50=p0*A^50
p100=p0*A^100
p0=[0 1 0]
p0*A^10
p0*A^50
clf
for n=1:5000
X=1;
for k=1:1000
X=Markov(X);
end
Y(n)=X;
end
plotrep(Y)
hold on
stairs(cumsum(P),'r')
hold off
%% 2
X=randn(1,1000);
Intconf(X,std(X),0.05)
k=0;
for i=1:20
    X=randn(1,1000);
Ic=Intconf(X,std(X),0.05)
if (0>Ic(1) && 0<Ic(2)) k=k+1;
end
end
k % k suit a peu pres une loi  binomiale de parametre (20,0.9)
%%
X=randn(1,1000);
a=[0.01:0.01:0.99];
m=length(a);
I1=zeros(1,m);
I2=zeros(1,m);
for i=1:m
    Ic=Intconf(X,std(X),a(i));
    I1(i)=Ic(1);
    I2(i)=Ic(2);
end
clf
plot(a,I1);
hold on
plot(a,I2);
%% 3
Intconfp(100,64,0.05)
Intconfp(400,212,0.05)

sondage(400,212)
sondage(100.52)
sondage(100,60)

File: 2012_2013_sujet_4_problemes_divers/TP4_2012_CP.py
"""TP4 2012-2013 Correction — Markov, IC, sondage."""
import numpy as np
import matplotlib.pyplot as plt

from plotrep import plotrep
from Markov import Markov
from Intconf import Intconf
from Intconfp import Intconfp
from sondage import sondage

# %% 1. Chaîne de Markov — loi stationnaire
A = np.array([[0, 1/3, 2/3], [1/3, 0, 2/3], [2/3, 1/3, 0]])
P_vec = np.array([0.35, 0.25, 0.4])
print(P_vec @ A)

p0 = np.array([0.05, 0.9, 0.05])
for n in [5, 10, 50, 100]:
    print(f'p{n} =', p0 @ np.linalg.matrix_power(A, n))

p0 = np.array([0, 1, 0])
print(p0 @ np.linalg.matrix_power(A, 10))
print(p0 @ np.linalg.matrix_power(A, 50))

Y = np.zeros(5000, dtype=int)
for n in range(5000):
    X = 1
    for k in range(1000):
        X = Markov(X)
    Y[n] = X
plt.figure()
plotrep(Y)
plt.step(np.arange(1, 4), np.cumsum(P_vec), 'r', where='post')
plt.title('Loi stationnaire chaîne de Markov')
plt.show()

# %% 2. Intervalles de confiance
X = np.random.randn(1000)
print('IC (sigma inconnu):', Intconf(X, np.std(X, ddof=1), 0.05))

k = 0
for i in range(20):
    X = np.random.randn(1000)
    Ic = Intconf(X, np.std(X, ddof=1), 0.05)
    if 0 > Ic[0] and 0 < Ic[1]:
        k += 1
print(f'k = {k}  (suit approx. Binomiale(20, 0.9))')

X = np.random.randn(1000)
a_vec = np.arange(0.01, 1.0, 0.01)
I1 = np.zeros(len(a_vec))
I2 = np.zeros(len(a_vec))
for i, ai in enumerate(a_vec):
    Ic = Intconf(X, np.std(X, ddof=1), ai)
    I1[i] = Ic[0]; I2[i] = Ic[1]
plt.figure()
plt.plot(a_vec, I1); plt.plot(a_vec, I2)
plt.xlabel('alpha'); plt.title('IC en fonction de alpha')
plt.show()

# %% 3. Intervalles de confiance pour une proportion
print('IC proportion (100, 64):', Intconfp(100, 64, 0.05))
print('IC proportion (400, 212):', Intconfp(400, 212, 0.05))

print('sondage(400, 212):', sondage(400, 212))
print('sondage(100, 52):', sondage(100, 52))
print('sondage(100, 60):', sondage(100, 60))
